File: web/source/dblp/publications.js
// JS port of sources/dblp/publications.py. DBLP's public search
// endpoint sets CORS and returns JSON when we pass `format=json`.

import { fetchJson, sleep } from "../utils/http.js";
import { truncate } from "../utils/doc.js";

const API = "https://dblp.org/search/publ/api";
const DELAY_MS = 1000;
const PAGE_SIZE = 100;
const ARXIV_RE = /(\d{4}\.\d{4,5})/;

function canonicalUrl(info) {
  const ee = info.ee || "";
  if (ee.includes("arxiv.org")) {
    const m = ee.match(ARXIV_RE);
    if (m) return `https://arxiv.org/abs/${m[1]}`;
  }
  if (ee.includes("doi.org")) return ee;
  return info.url || ee;
}

export async function publications({
  author,
  maxResults = 200,
  existingUrls = null,
}) {
  const out = {};
  let offset = 0;
  while (offset < maxResults) {
    const url = `${API}?q=author%3A${encodeURIComponent(
      author,
    )}&h=${PAGE_SIZE}&f=${offset}&format=json`;
    let result;
    try {
      result = await fetchJson(url);
    } catch (err) {
      console.warn("[sync] DBLP error:", err);
      break;
    }
    const hits = ((result.result || {}).hits || {}).hit || [];
    if (!Array.isArray(hits) || hits.length === 0) break;

    if (existingUrls) {
      const newInPage = hits.filter((h) => {
        const c = canonicalUrl(h.info || {});
        return c && !existingUrls.has(c);
      }).length;
      if (newInPage === 0) break;
    }

    for (const hit of hits) {
      const info = hit.info || {};
      // Python uses `.rstrip(".")` which strips a RUN of trailing
      // periods, not just one. Some DBLP titles end with "...".
      const title = (info.title || "").replace(/\.+$/, "");
      if (!title) continue;
      const canonical = canonicalUrl(info);
      if (!canonical) continue;
      if (existingUrls && existingUrls.has(canonical)) continue;
      if (out[canonical]) continue;

      const year = info.year || "";
      const venue = info.venue || "";
      let authorsRaw = (info.authors || {}).author || [];
      if (!Array.isArray(authorsRaw)) authorsRaw = [authorsRaw];
      // Python: `a.get("text", a) if isinstance(a, dict) else str(a)`.
      // When the dict lacks a "text" key, Python falls back to the
      // dict itself (later stringified into something like `{id: …}`).
      // We mirror that — real DBLP records always have "text", but
      // matching the edge-case behaviour keeps the ports byte-identical.
      const authors = authorsRaw
        .slice(0, 5)
        .map((a) => {
          if (a && typeof a === "object") {
            return a.text !== undefined ? String(a.text) : String(a);
          }
          return String(a);
        })
        .join(", ");

      const parts = [];
      if (authors) parts.push(authors);
      if (venue) parts.push(venue);
      const summary = truncate(parts.join(". "), 250);

      out[canonical] = {
        title,
        summary,
        date: year ? `${year}-01-01` : "",
        tags: ["scholar"],
        source: "scholar",
      };
    }

    offset += hits.length;
    if (hits.length < PAGE_SIZE) break;
    await sleep(DELAY_MS);
  }
  return out;
}

File: sources/dblp/publications.py
"""
DBLP computer science bibliography fetcher.

Fetches publications from DBLP, the definitive CS publication database.
Uses canonical URLs (arXiv > DOI > DBLP page) to deduplicate with
Semantic Scholar and Google Scholar.

API docs: https://dblp.org/faq/How+to+use+the+dblp+search+API.html
"""

import json
import re
import time
import urllib.request
import xml.etree.ElementTree as ET

__all__ = ["Publications"]

_API = "https://dblp.org/search/publ/api"
_DELAY = 1.0


def _fetch_xml(url: str) -> ET.Element:
    req = urllib.request.Request(url, headers={"User-Agent": "Knowledge/1.0"})
    with urllib.request.urlopen(req, timeout=15) as resp:
        return ET.fromstring(resp.read())


def _canonical_url(info: dict) -> str:
    """Pick best URL: arXiv > DOI > DBLP page."""
    ee = info.get("ee", "")
    if "arxiv.org" in ee:
        m = re.search(r"(\d{4}\.\d{4,5})", ee)
        if m:
            return f"https://arxiv.org/abs/{m.group(1)}"
    if "doi.org" in ee:
        return ee
    return info.get("url", ee)


class Publications:
    """Fetch publications from DBLP by author name."""

    def __init__(self, author: str, max_results: int = 200):
        self.author = author
        self.max_results = max_results

    def __call__(self, existing_urls: set[str] | None = None) -> dict[str, dict]:
        print(f"    Fetching DBLP publications for '{self.author}'...")
        data: dict[str, dict] = {}
        offset = 0
        page_size = 100

        while offset < self.max_results:
            url = f"{_API}?q=author%3A{urllib.parse.quote(self.author)}&h={page_size}&f={offset}&format=json"

            try:
                req = urllib.request.Request(url, headers={"User-Agent": "Knowledge/1.0"})
                with urllib.request.urlopen(req, timeout=15) as resp:
                    result = json.loads(resp.read())
            except Exception as e:
                print(f"    DBLP API error: {e}")
                break

            hits = result.get("result", {}).get("hits", {}).get("hit", [])
            if not hits:
                break

            # Page-level early exit: DBLP returns by relevance, not pure date, so
            # don't hard-break on first all-known page. Instead stop when we've
            # seen two consecutive all-known pages.
            if existing_urls is not None:
                new_in_page = 0
                for h in hits:
                    info = h.get("info", {})
                    canonical = _canonical_url(info)
                    if canonical and canonical not in existing_urls:
                        new_in_page += 1
                if new_in_page == 0:
                    print(f"    No new publications on page offset={offset}, stopping early.")
                    break
                print(f"    Offset {offset}: {new_in_page} new publication(s).")

            for hit in hits:
                info = hit.get("info", {})
                title = info.get("title", "").rstrip(".")
                if not title:
                    continue

                canonical = _canonical_url(info)
                if not canonical:
                    continue
                if existing_urls and canonical in existing_urls:
                    continue
                if canonical in data:
                    continue

                year = info.get("year", "")
                # `venue` is normally a string but DBLP returns a list
                # for cross-referenced publications (workshops inside
                # bigger venues, common for prolific authors like
                # Bengio). Flatten to a single comma-separated string
                # so the downstream `". ".join` doesn't choke on a
                # nested list.
                venue_raw = info.get("venue", "")
                if isinstance(venue_raw, list):
                    venue = ", ".join(str(v) for v in venue_raw if v)
                else:
                    venue = str(venue_raw or "")
                authors_raw = info.get("authors", {}).get("author", [])
                if isinstance(authors_raw, dict):
                    authors_raw = [authors_raw]

                # Each author may be a string OR a {"text": str, "@pid": ...} dict.
                # Coerce defensively so any unexpected shape (list-of-strings,
                # nested dict) becomes a flat string.
                def _author_str(a):
                    if isinstance(a, dict):
                        t = a.get("text", "")
                        return str(t) if not isinstance(t, list) else ", ".join(str(x) for x in t)
                    return str(a)

                authors = ", ".join(_author_str(a) for a in authors_raw[:5])

                summary_parts = []
                if authors:
                    summary_parts.append(authors)
                if venue:
                    summary_parts.append(venue)
                summary = ". ".join(summary_parts)
                if len(summary) > 250:
                    summary = summary[:247] + "..."

                data[canonical] = {
                    "title": title,
                    "summary": summary,
                    "date": f"{year}-01-01" if year else "",
                    "tags": ["scholar"],
                }

            offset += len(hits)
            if len(hits) < page_size:
                break
            time.sleep(_DELAY)

        print(f"    {len(data)} publications (deduplicated)")
        return data
